New exercise

wikfeldt opened the pull request:
wip, the exercise needs to be simplified and clarified with a graphic

bast commented:
Tip: prefix pull requests which you do not to be merged yet with "[WIP]", then it is clear that this is work in progress and nobody will merge by accident. If this is your intention to not merge yet but discuss first.

wikfeldt commented:
yes indeed, changed it. would be useful to get feedback on this

bast commented:
I agree that figures will help since it takes 2 or 3 reads to really understand what is needed. The bare/nonbare creates additional complexity but I see the value in this exercise. The "create a bare repo" will be unfamiliar so either we show how, or we let people search, or we could let them directly clone the GitHub repo with `git clone --mirror` which will mirror the entire repo by automatically creating a bare repo. Then we don't have to explain or even mention bare/nonbare since everybody probably understands what mirroring means. The mirrored (bare) repo can then be cloned to two places and then one can exercise. Feel free to do it differently than I suggest here, though, this is just a personal perspective and perhaps it does not make anything simpler.

bast commented:
As a side note GitLab will refuse to merge "[WIP]" PRs. Would be nice to also have the same in GitHub but it is a convention normally understood by people.

wikfeldt commented:
yeah creating a bare repo probably needs further clarification, but bare vs non-bare is described in the beginning of the episode (including the command git init --bare)

bast commented:
That is true but we could consider removing that part. As Sabry pointed out some time ago, most people never create bare repositories. Either they create local non-bare ones or they use GitHub/Lab and don't care that they are actually bare. Possibly we would not lose anything by never mentioning these concepts. I am not sure. First time I had to worry about it was when I tried to set up a repository that I would use to sync between several machines. I wonder whether people do that these days. Well one participant yesterday asked about it.

wikfeldt commented:
great feedback, thanks. Perhaps git clone --mirror is the best way to simplify the exercise. the idea is to get used to thinking about remotes, and creating a bare repo from scratch isn't necessary for that. cloning a mirror is more similar to forking, so it creates a conceptual bridge to the next episode.
also, removing the bare/nonbare discussion from the beginning might make sense

bast commented:
Cool! I think the analogy of `git clone --mirror` and fork is good. Minus all the web stuff around that. The `git clone --mirror` is also a good way to manually move repositories from one service to another since it moves everything and you don't forget tags or branches.

wikfeldt commented:
i've switched to using `git clone --mirror`, and added graphics to help with the mental model of what's going on. I think this might be useful, both to get used to remotes and to practice resolving conflicts.

bast commented:
Oops there is a conflict now due to the other PR that I just merged - can you have a look please? Sorry for extra work. Otherwise this is IMO very good and ready to be merged.

wikfeldt commented:
thanks for the review. i wasn't sure how smart git would be about this, but it was a trivial merge
